### DISCUSSION TITLE
Conformer v0.1.0

### DISCUSSION THREAD

**Discussion by Rishit-dagli** [Rishit-dagli/Conformer#2] [Announcements]
This is the initial release of Conformer and implements the Conformer (Convolution-augmented Transformer for Speech Recognition) as a subclassed TensorFlow model. Conformer achieves the best of both worlds (transformers for content-based global interactions and CNNs to exploit local features) by studying how to combine convolution neural networks and transformers to model both local and global dependencies of an audio sequence in a parameter-efficient way.

## Classes

You would majorly want to follow the examples and use the `ConformerConvModule` or `ConformerBlock` classes.

- [Attention](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/attention.py): Implements attention mechanism as shown in the paper as a TensorFlow Keras layer
- [Swish](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L9): The Swish activation function originally proposed by Ramachandran et al. in [Searching for Activation Functions](https://openreview.net/forum?id=SkBYYyZRZ)
- [GLU](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L17): The GLU activation function originally proposed by Dauphin et al. in [Language Modeling with Gated Convolutional Networks](https://icml.cc/Conferences/2017/Schedule?showEvent=887)
- [DepthwiseLayer](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L27): A modified depthwise convolutional layer
- [Scale](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L45): Apply an operation and scale it by a certain value
- [PreNorm](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L55): Normalize the activations of the previous layer for each given example in a batch independently and apply some function to it, implemented as a TensorFlow Keras Layer
- [FeedForward](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L66): Create a FeedForward neural net with two Dense and 2 Dropout layers and Swish activation, implemented as a TensorFlow Keras Layer
- [BatchNorm](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L82): Modified Conditional batchnorm layer
- [ConformerConvModule](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L93): The class for a ConformerConvModule, the major contribution of this paper
- [ConformerBlock](https://github.com/Rishit-dagli/Conformer/blob/main/conformer_tf/conformer_tf.py#L131): A complete ConformerBlock put together

<hr /><em>This discussion was created from the release <a href='https://github.com/Rishit-dagli/Conformer/releases/tag/v0.1.0'>Conformer v0.1.0</a>.</em>
